### PR TITLE
[RELAY] Refactor type inference to use type solver

### PR DESCRIPTION
**Pull request by tqchen**
- IncompleteType-> type.h
- Refactor the solving logic into TypeSolver with its own test case  
    - Unification logic merged into TypeSolver
- Recover TypeInferencer, document the three-step logic
- Pass the original set of test-cases

## Separation of Logics of Modules in This PR
- Arena: memory allocator
- TypeSolver: solving logic
- TypeInferencer: build up the constraints into TypeSolver and tries to resolve

## Known Limitations
Did not work a lot on error reporting and it is now quite bad when checker goes wrong, will need more systematic error message on what is the operator, relation and calling expression as well as the error context.

Maybe we could overload operator<<(Span) into something that takes a context Expr, print it and pin down on the location.

The unification is limited and only handles the union case. I feel it is useful to defer the general unification to a separate PR, with concerns listed as below.

### CODE REVIEW

File: include/tvm/relay/type.h [hunk 1]
```diff
@@ -214,55 +332,37 @@ class TypeRelation;
  */
 class TypeRelationNode : public TypeConstraintNode {
  public:
-  /*! \brief The name of the function */
```
Comment (jroesch) [inline include/tvm/relay/type.h]: Could we have some kind of name field for debugging/printing purposes?

File: include/tvm/runtime/packed_func.h [hunk 2]
```diff
@@ -175,7 +175,7 @@ class TypedPackedFunc<R(Args...)> {
    *
    * \param packed The packed function
    */
-  inline explicit TypedPackedFunc(PackedFunc packed);
+  inline TypedPackedFunc(PackedFunc packed);  // NOLINT(*)
```
Comment (jroesch) [inline include/tvm/runtime/packed_func.h]: 👍

File: src/relay/ir/type.cc [hunk 3]
```diff
@@ -79,24 +99,28 @@ TVM_STATIC_IR_FUNCTOR_REGISTER(IRPrinter, vtable)
...
 });
 
 TVM_STATIC_IR_FUNCTOR_REGISTER(IRPrinter, vtable)
-.set_dispatch<TypeRelationNode>([](const TypeRelationNode *node,
-                                       tvm::IRPrinter *p) {
-  p->stream << "TypeRelationNode(" << node->name << ", " << node->args
-    << ")";
+.set_dispatch<TypeRelationNode>([](const TypeRelationNode *node, tvm::IRPrinter *p) {
+    p->stream << "TypeRelationNode(";
+    p->print(node->func);
```
Comment (jroesch) [inline src/relay/ir/type.cc]: We could use the `node->func->name` for debugging instead of a secondary name.

File: src/relay/pass/type_infer.cc [hunk 4]
```diff
@@ -22,607 +22,360 @@
...
-  TypeUnifier unifier;
-
-  template <typename T>
-  T WithScope(const std::function<T()>& f) {
-    TypeContext::Scope fr(context);
-    return f();
+  // Visitor logics
+  Type VisitExpr_(const VarNode* op) final {
+    // The type of Var can already been lookedup in type_map_;
+    LOG(FATAL) << "Cannot find binding for var " << GetRef<Var>(op);
```
Comment (jroesch) [inline src/relay/pass/type_infer.cc]: I feel like we should still keep an error reporting interface which is span focused. The problem here is that the span on the variable is meaningless due to it's underlying container being contained in many places in the expression.

File: src/relay/pass/type_infer.cc [hunk 5]
```diff
@@ -22,607 +22,360 @@
...
-
-  std::vector<Expr> field_exprs;
-  std::vector<Type> field_types;
-  for (auto field = pl->fields.begin(); field != pl->fields.end(); field++) {
-    auto checked_field = Infer(*field);
-    field_exprs.push_back(checked_field.expr);
-    field_types.push_back(checked_field.type);
+  Type VisitExpr_(const ParamNode* op) final {
+    // directly handled by Funtion
+    LOG(FATAL) << "not reached";
```
Comment (jroesch) [inline src/relay/pass/type_infer.cc]: We should probably create a platform specific unreachable macro, we did this in Lean and its good to inform the compiler that some cases should never be reached.

File: src/relay/pass/type_infer.cc [hunk 6]
```diff
@@ -22,607 +22,360 @@
...
-  return {ParamNode::make(param->var, rtype), rtype};
-}
-
-CheckedExpr TypeInferencer::VisitFunction(const Function& f, bool generalize) {
-  // First we add the parameters to the context allowing us to check their
-  // types.
-
-  // TODO(@jroesch): support polymorphism
+  Type VisitExpr_(const GlobalVarNode* op) final {
+    CHECK(env_.defined()) << "DO not have environment to lookup GlobalVar";
```
Comment (jroesch) [inline src/relay/pass/type_infer.cc]: This field should never be undefined?

File: src/relay/pass/type_solver.cc [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+  for (TypeNode* ptr : type_nodes_) {
+    ptr->~TypeNode();
+  }
+  for (RelationNode* ptr : rel_nodes_) {
+    ptr->~RelationNode();
+  }
+}
+
+// Add equality constraint
+Type TypeSolver::Unify(const Type& dst, const Type& src) {
```
Comment (jroesch) [inline src/relay/pass/type_solver.cc]: This doesn't actually implement first order unification on types but just the equivalent of the `Union` operation on the union find. In order for type inference to work we must implement all the functionality from the previous unifier. 

For example I don't see how this handles this case `(?A, ?B) -> ?C = (Tensor<(10, 10), ?bt>, ?W) -> ?Z`, in order to have inference work we need to be able to handle these cases, this is why we had the previous type unifier class.

**Comment by jroesch** [comment]
cc @slyubomirsky @MarisaKirisame please review

File: src/relay/pass/type_solver.h [hunk 8]
```diff
@@ -0,0 +1,225 @@
...
+        head = tail = node;
+      }
+    }
+  };
+  /*!
+   * \brief type node struct
+   *  TypeNode implements a union-find data structure
+   *  that can unifies the same types to the name resolved_type.
+   */
+  struct TypeNode {
```
Comment (jroesch) [inline src/relay/pass/type_solver.h]: I think it would be better to name this UnionFind/wrap it up.

File: include/tvm/relay/op.h [hunk 9]
```diff
@@ -348,23 +349,25 @@ inline OpRegistry& OpRegistry::add_argument(const std::string& name,
...
-
-  TypedEnvFunc<Array<Type>(const Array<Type>&, int)> env_type_rel_func;
+  TypeRelationFn env_type_rel_func;
 
   if (runtime::Registry::Get(func_name)) {
     auto env_func = EnvFunc::Get(func_name);
     env_type_rel_func = env_func;
   } else {
     runtime::Registry::Register(func_name)
-        .set_body_typed<Array<Type>(const Array<Type>&, int)>(type_rel_func);
```
Comment (jroesch) [inline include/tvm/relay/op.h]: Why not set the typed body here?

File: include/tvm/relay/op.h [hunk 10]
```diff
@@ -375,15 +378,21 @@ inline OpRegistry& OpRegistry::add_type_rel(
...
-  auto ty_call_args = Array<Type>(arg_types);
+  Array<Type> ty_call_args = arg_types;
 
   // Add output type.
   auto out_param = TypeParamNode::make("out", TypeParamNode::Kind::kType);
   type_params.push_back(out_param);
+  // this will trigger copy on write.
   ty_call_args.push_back(out_param);
 
+  // The attributes of primitive op is nullptr
```
Comment (jroesch) [inline include/tvm/relay/op.h]: Could you elaborate?

File: include/tvm/relay/type.h (hunk 1 above)
Comment (tqchen) [reply include/tvm/relay/type.h]: The name is already in the EnvFunc

**Comment by tqchen** [comment]
re: general Unification no longer holds. I intentionally leave that because I kind of do not like the previous way of handling general unification. 

Specifically:
-  I was hoping that when we try to Unify, ```Tuple(?A, ?B) == Tuple(?C, 1)``` it should return a new incomplete type ```?t```, who have a new relation ```?t = Tuple(?A, ? B)``` and ```?A ==?C ```
- The goal is to always make sure that the types in the system are either Incomplete, in which case it can be solved by relations or concrete and avoid cases where there is a composite type of incomplete ones

File: include/tvm/relay/op.h (hunk 9 above)
Comment (tqchen) [reply include/tvm/relay/op.h]: There is no type checking here, so just to type less type signatures..

**Comment by slyubomirsky** [comment]
Avoiding nested incomplete types sounds like a good way to reduce headaches

File: src/relay/pass/type_solver.cc (hunk 7 above)
Comment (slyubomirsky) [reply src/relay/pass/type_solver.cc]: Could the case you gave as an example be handled by going down the AST of the function type and generating equality constraints for each branch? Should that be an updated version of the current unifier?

**Comment by jroesch** [comment]
Type inference requires that we unify partial types, without it we don't have type inference. I feel like we keep trying to change the algorithm in ways which changes it from something that is not inference.

The whole point of doing type inference is that the constraints which cause a type to be complete are scattered throughout the program and won't manifest at a single use site, and are recursive by nature.

The reason type inference algorithms use union find is to manage the number of constraints generated and decompose them into the minimal number of primitive constraints between types, which may further give rise to other constraints. 

The first solution is exactly what happens under the previous algorithm, except yours introduces an unneeded variable, we don't need a fresh `?t` at all anything that unifies with `Tuple(?A, ?B)` must be `Tuple(_, _)` and this transitively holds even with a fresh variable.

Even when you have a complete inference algorithm the solution may be under constrained and you  can end up with a composite type containing incomplete types, if you are not OK with this then you don't want type inference.

For example there is a bug in the current inferencer which can not be resolved without supporting nested incomplete types. 

When checking a call we must unify the type of the node with a function type of the form `(T_1, ..., T_N) -> R` where `N` is the number of arguments, without this behavior we can't support polymorphism where there is an outer constructor wrapping the type argument.

File: src/relay/pass/type_solver.cc (hunk 7 above)
Comment (slyubomirsky) [reply src/relay/pass/type_solver.cc]: Oh I see the point, owing to your comment on the main thread: The example you gave requires nested incomplete types and is impossible to give without nested incomplete types

**Comment by tqchen** [comment]
To be clear, I am all for supporting Unifying the composite types with incomplete types. But just think a second PR is necessary to sort that out.

My main point is to introduce a fresh incomplete type ```?t``` for handling composite types. With the following reasons:

- The current algorithm tries to implement a fast version of inference that does lazy triggers of the active relations in the bipartite graph
- Specifically, when a Type get updated by the solver, all the unsolved relations that refer to the Type will get triggered.
- However, to make sure such relation propagates nicely, we will need to allocate a node for each type that is incomplete, say a relation which relies on a Tuple(?A, ? B), which is compositional. If there is no node in the bipartite graph to represent this type, then the triggering is flawed
   - Conceptually, when a composite type contains incomplete holes, we can also view it as a constraint between ?T and ?A, ?B if we generate a fresh ```?T```
- The main concern I have is when the inference gets ties into symbolic integers, in this case, actually, the symbolic integer itself can be incomplete, which means TensorType is a composite type instead of a primitive one, having a strict rule such that incomplete type composition is always represented by relation might help mitigate this issue.

Given these thoughts, I feel it is better to bring a second PR to solve these issues. Note that the original way of treating incomplete composite type solves the issue of composite tuples but not the symbolic integer case. I hope we can check in things that are clean, limited but known to work on limited cases, then improve them to cover a broader set of things that we care about.

**Comment by tqchen** [comment]
@jroesch I tried to address most of your comments. I feel the code could need another small refactor once we introduced a span based reporting with real error message model. I have added comments to state that the current version intentionally left out compositional type resolving for a future PR. 

The current part is sufficient for moving forward with simple type checking, while leaving room for future enhancements to TypeSolver without changing the interface

**Comment by tqchen** [comment]
@junrushao1994 can you also help review this?

File: include/tvm/relay/op.h [hunk 11]
```diff
@@ -375,15 +378,28 @@ inline OpRegistry& OpRegistry::add_type_rel(
...
+  //
+  // The attributes of primitive operator can vary at the call site.
+  // The type of sum is also dependent on Attrs being passed.
+  // So puting nullptr in the Attrs a special meaning,
```
Comment (MarisaKirisame) [inline include/tvm/relay/op.h]: has a special meaning

File: src/relay/pass/type_solver.cc [hunk 12]
```diff
@@ -0,0 +1,166 @@
...
+  TypeNode* lhs = GetTypeNode(dst);
+  TypeNode* rhs = GetTypeNode(src);
+  if (lhs->resolved_type.as<IncompleteTypeNode>()) {
+    MergeFromTo(lhs, rhs);
+    return rhs->resolved_type;
+  } else if (rhs->resolved_type.as<IncompleteTypeNode>()) {
+    MergeFromTo(rhs, lhs);
+    return lhs->resolved_type;
+  } else {
+    lhs->parent = rhs;
```
Comment (MarisaKirisame) [inline src/relay/pass/type_solver.cc]: what if it is not nullptr? maybe call find and lhs->parent->parent = rhs?

Comment (tqchen) [reply src/relay/pass/type_solver.cc]: We already call FindRoot in the beginning of the logic so lhs and rhs's parent is always nullptr

Comment (MarisaKirisame) [reply src/relay/pass/type_solver.cc]: can we put an assert (or a comment stating the invariant? it is not obvious.)

**Comment by MarisaKirisame** [comment]
I think there will be such 'special case' for every type constructor (reference, tuple, function, tensor).
Can we see the solver as having two separate part: a traditional algol-w union find, and a type relation solver?
While it seems more ad hoc, IMO adding a bunch of special case is worse.
Another reason is we probably need a fancier type inference algorithm, as we are system f, but is currently using algol w - it only work to hindley milner, so first class forall (basically virtual template) cannot get typed. Or we should change our doc to clarify that we are hm.

**Comment by tqchen** [comment]
This is merged for now, Thanks @jroesch @MarisaKirisame @slyubomirsky !

**Pull request merged**